storage: refactor log truncation index computation

tbg opened the pull request:
This makes it a lot easier to log descriptive debug messages indicating
how a truncation decision was arrived at, and in particular allows
pointing the finger at truncations that lead to Raft snapshots, which
is relevant in the context of #32046.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32137)
<!-- Reviewable:end -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32137)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/raft_log_queue.go, line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo0Mlb3r8-v29U2pHk:-LQo0Mlb3r8-v29U2pHl:barwk8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L155)):*
> ```Go
> )
> 
> type truncateDecisionInput struct {
> ```

Nit: I'm not aware of other usage of the term `Input` like this. Perhaps `Args` instead.
___
*[pkg/storage/raft_log_queue.go, line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo12K59RH5IdP4FYwy:-LQo12K59RH5IdP4FYwz:b-gh9lsl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L170)):*
> ```Go
> }
> 
> func (td *truncateDecision) NumResultingRaftSnapshots() int {
> ```

Nit: `Resulting` is a tad confusing. Perhaps `NumNewRaftSnapshots` as this method is computing the number of new Raft snapshots that will be caused by the truncation.

Does defining this method on a pointer receiver cause the receive to be pushed to the heap? Using a non-pointer receiver might not be good either as `truncateDecision` isn't small.
___
*[pkg/storage/raft_log_queue.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo0lty3BG4QHK4hsrz:-LQo0ltzDBpUfekTTatm:b-5xwh7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L187)):*
> ```Go
> 	}
> 
> 	return numCutOffAfter - numCutOffBefore
> ```

I was mildly confused by what this method was doing at first glance. I think it could be simplified and made clearer by adding a `raftSnapshotsForIndex` method. The method here would become:

```
  return td.raftSnapshotsForIndex(td.NewFirstIndex) - td.raftSnapshotsForIndex(td.Input.FirstIndex)
```
___
*[pkg/storage/raft_log_queue.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo-cw7EJpiNCBrSHzy:-LQo-cw7EJpiNCBrSHzz:b-7o0k27) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L206)):*
> ```Go
> func (td *truncateDecision) NumTruncatableIndexes() int {
> 	if td.NewFirstIndex < td.Input.FirstIndex {
> 		log.Fatalf(context.TODO(), "invalid truncate decision: first index would move from %d to %d", td.Input.FirstIndex, td.NewFirstIndex)
> ```

Nit: Are we ever going to plumb a context here? I'd make this `context.Background()`. Also, that line should be wrapped.
___
*[pkg/storage/raft_log_queue.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo1v1tCSGyXBgEpe2M:-LQo1v1tCSGyXBgEpe2N:b8ocdlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L336)):*
> ```Go
> 	// Can and should the raft logs be truncated?
> 	if decision.ShouldTruncate() {
> 		if n := decision.NumResultingRaftSnapshots(); n > 0 && rlq.logSnapshots.ShouldProcess(timeutil.Now()) {
> ```

Should the condition above include `log.V(1) || ...`. 


<!-- Sent from Reviewable.io -->

nvb reviewed (dismissed):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32137)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/raft_log_queue.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6uWY18QrLQbgAiE5Y:-LR6uWY18QrLQbgAiE5Z:b-krg4ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L89)):*
> ```Go
> // getTruncatableIndexes returns a truncateDecision for the given Replica. If no
> // truncation can be carried out, a zero decision is returned.
> func getTruncatableIndexes(ctx context.Context, r *Replica) (truncateDecision, error) {
> ```

nit: This should probably be renamed to something like `makeTruncationDecision` or `decideOnTruncation` now that it's no longer returning a set of indices.
___
*[pkg/storage/raft_log_queue.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6lk3k5Dr10b0c0p2C:-LR6lk3k5Dr10b0c0p2D:b-8y1uf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L92)):*
> ```Go
> 	rangeID := r.RangeID
> 
> 	r.mu.RLock()
> ```

Why lock twice? Shouldn't we push this into the lock directly below? Also, can that use an `RLock`?
___
*[pkg/storage/raft_log_queue.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6tfYN5TePkSqXlo7O:-LR6tfYN5TePkSqXlo7P:b-3sonib) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L159)):*
> ```Go
> 	LogSize, MaxLogSize   int64
> 	FirstIndex, LastIndex uint64
> 	PendingSnapshotIndex  uint64
> ```

Any way I can convince you to rename all references of this to `PendingPreemptiveSnapshotIndex` (or something similar)?
___
*[pkg/storage/raft_log_queue.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6tb6k0Bsu8EOhwfLi:-LR6tb6k0Bsu8EOhwfLj:blvzccq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L180)):*
> ```
> td.Input.PendingSnapshotIndex < td.NewFirstIndex
> ```

Should this be `td.Input.PendingSnapshotIndex < td.Input.FirstIndex`?
___
*[pkg/storage/raft_log_queue.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo0lty3BG4QHK4hsrz:-LR6toGI1CFeXwXmCCNz:b2k3521) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L187)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I was mildly confused by what this method was doing at first glance. I think it could be simplified and made clearer by adding a `raftSnapshotsForIndex` method. The method here would become:

```
  return td.raftSnapshotsForIndex(td.NewFirstIndex) - td.raftSnapshotsForIndex(td.Input.FirstIndex)
```
</blockquote></details>

+1 that would be much clearer.
___
*[pkg/storage/raft_log_queue.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6tqTy5HAkvFtCUyR_:-LR6tqTy5HAkvFtCUyRa:b705aam) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L191)):*
> ```Go
> 
> func (td *truncateDecision) String() string {
> 	var buf bytes.Buffer
> ```

nit: use `strings.Builder`


<!-- Sent from Reviewable.io -->

tbg committed:
storage: refactor log truncation index computation

This makes it a lot easier to log descriptive debug messages indicating
how a truncation decision was arrived at, and in particular allows
pointing the finger at truncations that lead to Raft snapshots, which
is relevant in the context of #32046.

Release note: None

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32137)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/raft_log_queue.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6lk3k5Dr10b0c0p2C:-LR70zlA-4_wjebfW2Eq:b-v6jm3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L92)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why lock twice? Shouldn't we push this into the lock directly below? Also, can that use an `RLock`?
</blockquote></details>

It can't use an `RLock` because `raftFirstIndexLocked` needs the exclusive lock (it modifies the cached truncated state). Since you don't seem concerned about running the larger `Lock` section unconditionally, I've made that change.
___
*[pkg/storage/raft_log_queue.go, line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo0Mlb3r8-v29U2pHk:-LR6xoDG8cluISdVG9Ln:b8nh5x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L155)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: I'm not aware of other usage of the term `Input` like this. Perhaps `Args` instead.
</blockquote></details>

I don't feel strongly about this, but please consider that we use "Args" in the context of our batch requests (we also have `TruncateLogRequest` which is also dubbed `TruncateLogArgs`), so the rename seems to add confusion rather than remove it. Let me know if you still prefer it.
___
*[pkg/storage/raft_log_queue.go, line 159 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6tfYN5TePkSqXlo7O:-LR6y2hQ8YOwqk3jFNPH:bjudzxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L159)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Any way I can convince you to rename all references of this to `PendingPreemptiveSnapshotIndex` (or something similar)?
</blockquote></details>

I actually have a WIP that makes this unconditional, but I'll rename it in the meantime.
Done.
___
*[pkg/storage/raft_log_queue.go, line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo12K59RH5IdP4FYwy:-LR71vbT0aEjxa0g-Ex0:b3okgjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L170)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: `Resulting` is a tad confusing. Perhaps `NumNewRaftSnapshots` as this method is computing the number of new Raft snapshots that will be caused by the truncation.

Does defining this method on a pointer receiver cause the receive to be pushed to the heap? Using a non-pointer receiver might not be good either as `truncateDecision` isn't small.
</blockquote></details>

Done. I'm also returning a pointer now.
___
*[pkg/storage/raft_log_queue.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LR6tb6k0Bsu8EOhwfLi:-LR6yO7D3qqc8EZ4y4SL:b-t42m2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L180)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> td.Input.PendingSnapshotIndex < td.NewFirstIndex
> ```

Should this be `td.Input.PendingSnapshotIndex < td.Input.FirstIndex`?
</blockquote></details>

Yep. I found out the hard way when log messages weren't showing up in my experiments. Testing catches this now, too.
___
*[pkg/storage/raft_log_queue.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo0lty3BG4QHK4hsrz:-LR6yvAE9uD66ZvTAJig:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L187)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1 that would be much clearer.
</blockquote></details>

Good idea, done.
___
*[pkg/storage/raft_log_queue.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo1v1tCSGyXBgEpe2M:-LR6zMO31uuBsEwqguwg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L336)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should the condition above include `log.V(1) || ...`. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32137)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/raft_log_queue.go, line 155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32137#-LQo0Mlb3r8-v29U2pHk:-LR7JFlg7q_9PxJEoG7M:b-6shv60) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeffbb59fc705b97d25f548b558122db5485dd4/pkg/storage/raft_log_queue.go#L155)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I don't feel strongly about this, but please consider that we use "Args" in the context of our batch requests (we also have `TruncateLogRequest` which is also dubbed `TruncateLogArgs`), so the rename seems to add confusion rather than remove it. Let me know if you still prefer it.
</blockquote></details>

Heh, the symmetry with the existing `Args` usage was why I suggested `Args`. I don't feel strongly about this, though.


<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

bors r=petermattis
